universe/NamedValueRefManager.cpp

```diff
@@ -0,0 +1,190 @@
+    return nullptr;
+}
+
+// default implementation - queries the untyped registry
+template <typename T>
+ValueRef::ValueRef<T>* const NamedValueRefManager::GetValueRef(const std::string& name) /*const*/ {
+    return dynamic_cast<ValueRef::ValueRef<T>*>(this->GetValueRefImpl(m_value_refs, "generic", name));
+}
+
+// int specialisation - queries the ValueRef<int> registry
+template <>
+ValueRef::ValueRef<int>* const NamedValueRefManager::GetValueRef(const std::string& name) /*const*/ {
+    return this->GetValueRefImpl(m_value_refs_int, "int", name);
+}
+
+// double specialisation - queries the ValueRef<double> registry
+template <>
+ValueRef::ValueRef<double>* const NamedValueRefManager::GetValueRef(const std::string& name) /*const*/ {
+    return this->GetValueRefImpl(m_value_refs_double, "double", name);
+}
+
+ValueRef::ValueRefBase* const NamedValueRefManager::GetValueRefBase(const std::string& name) const {
+    /* TODO straighten out const shtuff */
+    auto* drefp = const_cast<NamedValueRefManager*>(this)->GetValueRef<double>(name);
+    //if (auto* drefp = const_cast<NamedValueRefManager*>(this)->GetValueRef<double>(name)) // TODO C++17
+    if (drefp)
+        return drefp;
+    auto* irefp = const_cast<NamedValueRefManager*>(this)->GetValueRef<int>(name);
+    //if (auto* irefp = const_cast<NamedValueRefManager*>(this)->GetValueRef<int>(name)) // TODO C++17
+    if (irefp)
+        return irefp;
+    const auto it = m_value_refs.find(name);
+    return it != m_value_refs.end() ? it->second.get() : nullptr;
+}
+
+NamedValueRefManager& NamedValueRefManager::GetNamedValueRefManager() {
+    DebugLogger() << "NamedValueRefManager::GetNamedValueRefManager starts (check the thread)";
+    
+    static NamedValueRefManager manager; // function local 
+    InfoLogger() << "NamedValueRefManager::GetNamedValueRefManager at " << &manager;
+    return manager;
+}
+
+
+unsigned int NamedValueRefManager::GetCheckSum() const {
+    unsigned int retval{0};
+    for (auto const& name_type_pair : m_value_refs)
+        CheckSums::CheckSumCombine(retval, name_type_pair);
+    for (auto const& name_type_pair : m_value_refs_int)
+        CheckSums::CheckSumCombine(retval, name_type_pair);
+    for (auto const& name_type_pair : m_value_refs_double)
+        CheckSums::CheckSumCombine(retval, name_type_pair);
+    //CheckSums::CheckSumCombine(retval, m_value_refs.size()); // why also add size?
+    
+    DebugLogger() << "NamedValueRefManager checksum: " << retval;
+    return retval;
+}
+
+
+NamedValueRefManager::any_container_type  NamedValueRefManager::GetItems() const {
+    std::function<const any_entry_type(const entry_type&)> base_to_base          = [](const entry_type& kv) { return any_entry_type(kv.first, *(kv.second.get())); };
+    std::function<const any_entry_type(const double_entry_type&)> double_to_base = [](const double_entry_type& kv) { return any_entry_type(kv.first, *(kv.second.get())); };
```

Comment:
could put the lambda into the `std::transform` call and avoid storing it in a `std::function` ?

Reply:
yes, it makes the transform calls very long (lines about 200 characters) and the name expressing the intention is lost.
does it make any runtime difference not storing it? i will push a more concise version without assignment

Reply:
you can break the transform call and lambda over a few lines

not much if any runtime difference probably